python/cudf/cudf/tests/test_scalar.py

```diff
@@ -289,3 +289,14 @@ def test_device_scalar_direct_construction(value):
         assert s.dtype == "object"
     else:
         assert s.dtype == dtype
+
+
+@pytest.mark.parametrize("value", SCALAR_VALUES)
+def test_construct_from_scalar(value):
+    value = cudf.utils.utils.to_cudf_compatible_scalar(value)
+    x = cudf.Scalar(1, value.dtype)
+    y = cudf.Scalar(x)
+    assert x.value == y.value
+
+    # check that this works:
+    y.device_value
```

Comment:
Should we test that the synchronization state of the result is the same as the original?

```
x._is_host_value_current == y._is_host_value_current
x._is_device_value_current == y._is_device_value_current
```

Reply:
I'd be -1 to testing internals like this since they are implementation details that can easily change. But  it looks like we're doing it already in these tests. It's your call: I'm happy to add it on

Reply:
One can argue that we should have tests that test only API functionality - just checking that the results of calling our methods and functions are what we expect them to be. 

Then there's things that we actually want to work a certain way under the hood, that can easily do something unexpected or unwanted when subject to changes - I'd argue we want to test that too. It's just that in lieu of doing it here, I am not sure where else to put them.

Reply:
I definitely see your point, and have added the test as asked.

I think in this specific case, the only external/user-facing behaviour is performance, where ideally that would be captured in a _benchmark_ rather than a _test_. 

The problem with testing internals/optimizations is that when we _do_ decide to change implementation details, we have to delete tests. At that point, there may be difficult-to-answer questions like:

* should we replace the tests we are deleting with new tests?
* will deleting these tests now cause some user-facing behaviour to be untested?
* who can decide?